tools/shared/src/main/scala/org/scalajs/core/tools/javascript/ScalaJSClassEmitter.scala

```diff
@@ -39,33 +40,24 @@ final class ScalaJSClassEmitter private (
     this(semantics, outputMode, linkingUnit, linkingUnit.globalInfo)
   }
 
-  @deprecated(
-      "This constructor creates an emitter that cannot handle JS classes. " +
-      "Use the constructor with a LinkingUnit instead.", "0.6.5")
-  def this(semantics: Semantics, outputMode: OutputMode,
-      globalInfo: LinkingUnit.GlobalInfo) =
-    this(semantics, OutputMode.ECMAScript51Global, null, globalInfo)
-
-  @deprecated(
-      "This constructor creates an emitter that cannot handle JS classes. " +
-      "Use the constructor with a LinkingUnit instead.", "0.6.2")
-  def this(semantics: Semantics, globalInfo: LinkingUnit.GlobalInfo) =
-    this(semantics, OutputMode.ECMAScript51Global, globalInfo)
-
-  @deprecated(
-      "This constructor creates an emitter that cannot handle JS classes. " +
-      "Use the constructor with a LinkingUnit instead.", "0.6.1")
-  def this(semantics: Semantics) =
-    this(semantics, LinkingUnit.GlobalInfo.SafeApproximation)
-
   private[javascript] lazy val linkedClassByName: Map[String, LinkedClass] = {
-    if (linkingUnit == null) {
-      throw new IllegalArgumentException(
-          "A class emitter created without a LinkingUnit cannot emit JS classes")
-    }
     linkingUnit.classDefs.map(c => c.encodedName -> c).toMap
   }
 
+  lazy val usesJSConstructorOpt: Set[String] = {
+    val blackList = Set("jl_ClassCastException",
+      "sjsr_UndefinedBehaviorError", "jl_NullPointerException",
+      "js_CloneNotSupportedException")
+    val defsNoInst = linkingUnit.classDefs.filter(_.hasInstances)
+    val cantCtorOpt = (defsNoInst.flatMap(_.superClass) ++
+        defsNoInst.filterNot(_.classExports.isEmpty).
+        map(_.name)).map(_.name).toSet ++ blackList
```

Comment:
Can you make this more readable? Maybe move the filter on class exports to the second part (the "positive" filter)?